internal/app/go-filecoin/retrieval_market_connector/client.go

```diff
@@ -1,27 +1,292 @@
+	// GetDefaultWalletAddress retrieves the wallet addressed used to sign data and pay fees
+	GetDefaultWalletAddress() (address.Address, error)
+}
 
-func NewRetrievalClientNodeConnector() *RetrievalClientNodeConnector {
-	return &RetrievalClientNodeConnector{}
+type PaymentBrokerAPI interface {
+	// GetPaymentChannelID queries for the address of a payment channel for a payer/payee
+	GetPaymentChannelID(ctx context.Context, payer, payee address.Address) (address.Address, error)
+	// GetClientAddrByChannelID returns the PaymentChannel struct for a given payment channel
+	GetPaymentChannelInfo(ctx context.Context, paymentChannel address.Address) (address.Address, paymentbroker.PaymentChannel, error)
 }
 
-func (r *RetrievalClientNodeConnector) GetOrCreatePaymentChannel(ctx context.Context, clientAddress address.Address, minerAddress address.Address, clientFundsAvailable tokenamount.TokenAmount) (address.Address, error) {
-	panic("TODO: go-fil-markets integration")
+type RetrievalSigner interface {
+	SignBytes(data []byte, addr address.Address) (types.Signature, error)
 }
 
+func NewRetrievalClientNodeConnector(
+	bs blockstore.Blockstore,
+	cs *chain.Store,
+	mw MsgWaiter,
+	ob MsgSender,
+	ps piecestore.PieceStore,
+	sm SmAPI,
+	signer RetrievalSigner,
+	aapi ActorAPI,
+	wal WalletAPI,
+	pbapi PaymentBrokerAPI,
+) *RetrievalClientNodeConnector {
+	return &RetrievalClientNodeConnector{
+		bs:         bs,
+		cs:         cs,
+		mw:         mw,
+		outbox:     ob,
+		ps:         ps,
+		sm:         sm,
+		signer:     signer,
+		actAPI:     aapi,
+		wal:        wal,
+		pbAPI:      pbapi,
+		laneStore:  make(map[address.Address]laneEntries),
+	}
+}
+
+// GetOrCreatePaymentChannel gets or creates a payment channel and posts to chain
+// Assumes GetOrCreatePaymentChannel is called before AllocateLane
+// Blocks until message is mined?
+func (r *RetrievalClientNodeConnector) GetOrCreatePaymentChannel(ctx context.Context, clientWallet address.Address, minerWallet address.Address, clientFundsAvailable tokenamount.TokenAmount) (address.Address, error) {
+
+	if clientWallet == address.Undef || minerWallet == address.Undef {
+		return address.Undef, errors.New("empty address")
+	}
+
+	chid, err := r.pbAPI.GetPaymentChannelID(ctx, clientWallet, minerWallet)
+	if err != nil {
+		return address.Undef, err
+	}
+
+	if chid == address.Undef {
+		// create the payment channel
+		bal, err := r.wal.GetBalance(ctx, clientWallet)
+		if err != nil {
+			return address.Undef, err
+		}
+
+		filAmt := types.NewAttoFIL(clientFundsAvailable.Int)
+		if bal.LessThan(filAmt) {
+			return address.Undef, errors.New("not enough funds in wallet")
+		}
+
+		height, err := r.getBlockHeight()
+		if err != nil {
+			return address.Undef, err
+		}
+		validAt := height + 1 // valid almost immediately since a retrieval could theoretically happen in 1 block
+
+		to, err := GoAddrFromFcAddr(fcaddr.LegacyPaymentBrokerAddress)
```

Comment:
This function should go away now that go-address is fully integrated in filecoin.